packages/reactive-element/src/decorators/query.ts

```diff
@@ -64,10 +84,24 @@ export function query(selector: string, cache?: boolean): QueryDecorator {
     descriptor?: PropertyDescriptor
   ) => {
     const doQuery = (el: Interface<ReactiveElement>): V => {
+      const result = (el.renderRoot?.querySelector(selector) ?? null) as V;
+      if (DEV_MODE && result === null && cache && !el.hasUpdated) {
+        const name =
+          typeof nameOrContext === 'object'
+            ? nameOrContext.name
+            : nameOrContext;
+        issueWarning(
+          '',
+          `@query'd field ${JSON.stringify(String(name))} for selector ` +
+            `'${selector}' has been accessed before the first update. This ` +
+            `yields a certain null result if the renderRoot tree has not ` +
+            `been provided beforehand (e.g. via Declarative Shadow DOM).`
```

Comment:
Thanks for bearing with these changed, I think the last change required is to update the warning to relate to the situation that was handled. E.g., You are preventing the caching of a `null` value.

Reply:
I was just wondering if the message should closely reflect the conditions for issuing the warning 👍 
I've kept a brief explanation of the cause since there isn't a detailed explanation in the documentation yet. I think it could be removed later to keep the message concise.